ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -590,19 +627,29 @@ func (t *TPUWebhookServer) mutatePod(admissionReview *admissionv1.AdmissionRevie
 	chipsPerHost := getNumTPUChipsRequested(containers...)
 	numOfHosts, _ := getNumTPUHostsFromTopology(clusterName, groupName, namespace, topology, chipsPerHost) // ignore error here because topology may not be set yet
 
-	podsInGroup, err := t.podLister.Pods(namespace).List(labels.SelectorFromSet(labels.Set{"ray.io/group": groupName}))
-	if err != nil {
-		return nil, err
+	// Wait for PodInformer cache to update from previous requests or timeout
+	if waitTimeout(&t.wg, time.Second*1) {
+		klog.V(1).Info("MutatePod", "PodInformer AddFunc called for prior admission request")
+	} else {
+		klog.V(1).Info("MutatePod", "Timed out waiting for PodInformer AddFunc")
 	}
+	// Add 1 to the WaitGroup to represent the pending Pod to the cache
+	defer t.wg.Add(1)
+	t.waiting += 1
 
 	// query k8s client to populate sliceToWorkerIDs to then calculate the next TPU_WORKER_ID and replicaIndex
-	sliceToWorkerIDs, err := getSliceToWorkerIDs(podsInGroup, clusterName, groupName, namespace, numOfHosts)
+	sliceToWorkerIDs, err := t.getSliceToWorkerIDs(clusterName, groupName, namespace, numOfHosts)
 	if err != nil {
 		return nil, err
 	}
 	replicaIndex := getReplicaIndex(sliceToWorkerIDs, clusterName, groupName, namespace)
 	podSlice := slice{clusterName, groupName, namespace, replicaIndex, numOfHosts}
-	tpuWorkerID := getNextWorkerID(sliceToWorkerIDs, podSlice, namespace, replicaIndex) // defaults to 0 for single-host
+	tpuWorkerID, err := getNextWorkerID(sliceToWorkerIDs, podSlice, namespace, replicaIndex) // defaults to 0 for single-host
+	if err != nil {
+		return nil, err
+	}
+	// set the unique identifier for the last admitted Pod by this TPUWebhookServer
+	t.lastAdmitted = fmt.Sprintf("%s-%s-%d-%d", namespace, clusterName, replicaIndex, tpuWorkerID)
```

Comment:
Should last admitted be tracked per RayCluster?

Reply:
Since the Pods are admitted in series by each webhook replica I think it works the same to just check that the last admitted Pod has been added to the cache, regardless of RayCluster. If we did it per-RayCluster we'd also have to handle RayCluster deletion to clean up the list of `lastAdmitted` Pods.